### ISSUE TITLE
Fails to parse ?& array['foo']

### ISSUE THREAD

**Issue by jpmckinney** [sqlfluff/sqlfluff#6321]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Write test.sql:

```sql
SELECT '["a", "b", "c"]'::jsonb ?& array['a', 'b'];
```

Run:

```shell-session
$ sqlfluff lint
== [test.sql] FAIL                                                                                                                                                                                                 
L:   1 | P:  33 |  PRS | Line 1, Position 33: Found unparsable section: "?&
                       | array['a', 'b']"
WARNING: Parsing errors found and dialect is set to 'postgres'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

### Expected Behaviour

It parses this JSON expression, taken from https://www.postgresql.org/docs/current/functions-json.html

### Observed Behaviour

See What Happened

### How to reproduce

See What Happened

### Dialect

postgres

### Version

sqlfluff, version 3.2.2

### Configuration

[tool.sqlfluff.core]
dialect = "postgres"

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
